File: samples/client/3_1_0_unit_test/java/src/main/java/org/openapijsonschematools/client/schemas/validation/JsonSchemaFactory.java
package org.openapijsonschematools.client.schemas.validation;

import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;


public class JsonSchemaFactory {
    static Map<Class<? extends JsonSchema<?>>, JsonSchema<?>> classToInstance = new HashMap<>();

    public static <V extends JsonSchema<?>> V getInstance(Class<V> schemaCls) {
        @Nullable JsonSchema<?> cacheInst = classToInstance.get(schemaCls);
        if (cacheInst != null) {
            assert schemaCls.isInstance(cacheInst);
            return schemaCls.cast(cacheInst);
        }
        try {
            Method method = schemaCls.getMethod("getInstance");
            @SuppressWarnings("nullness") @NonNull Object obj = method.invoke(null);
            assert schemaCls.isInstance(obj);
            V inst = schemaCls.cast(obj);
            classToInstance.put(schemaCls, inst);
            return inst;
        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
            throw new RuntimeException(e);
        }
    }
}

File: samples/client/petstore/kotlin/src/main/kotlin/org/openapijsonschematools/client/schemas/validation/JsonSchemaFactory.kt
package org.openapijsonschematools.client.schemas.validation

import java.lang.reflect.InvocationTargetException

object JsonSchemaFactory {
    private var classToInstance: MutableMap<Class<out JsonSchema<*>>, JsonSchema<*>> = HashMap()

    fun <V : JsonSchema<*>> getInstance(schemaCls: Class<V>): V {
        val cacheInst: JsonSchema<*>? = classToInstance[schemaCls]
        cacheInst?.let {
            assert(schemaCls.isInstance(it))
            return schemaCls.cast(it)
        }
        try {
            val method = schemaCls.getMethod("getInstance")
            val obj = method.invoke(null)
            assert(schemaCls.isInstance(obj))
            val inst = schemaCls.cast(obj)
            classToInstance[schemaCls] = inst
            return inst
        } catch (e: NoSuchMethodException) {
            throw RuntimeException(e)
        } catch (e: IllegalAccessException) {
            throw RuntimeException(e)
        } catch (e: InvocationTargetException) {
            throw RuntimeException(e)
        }
    }
}